runtime/src/accounts_index.rs

```diff
@@ -768,13 +774,20 @@ impl<T: 'static + Clone> AccountsIndex<T> {
             //  - The secondary index is never consulted as primary source of truth for gets/stores.
             //  So, what the accounts_index sees alone is sufficient as a source of truth for other non-scan
             //  account operations.
+            if account_info.is_zero_lamport() {
+                self.zero_lamport_accounts.insert(*pubkey);
+            }
```

Comment:
strictly spekaing, this new code must be inside `w_account_enty` lock? It can't be moved out to before getting `w_account_entry`?

Reply:
Well, that's a good question. I was thinking it needs to be in at least a read-lock on the entry. We don't take a read-lock on the update flow, so I'm not sure it makes sense to add another lock just to update this set.

It's not great to do more work in the write critical section, but I think zero-lamport store case is relatively un-common.

Reply:
If I do it above outside the lock, then there is a race between the clean thread and the insert thread.
1. insert thread inserts the key in to zero_lamport_keys, but doesn't get into the write lock
2. clean thread grabs the read lock on the entry
3. clean thread sees no zero-lamport updates in the list and thus removes key from zero_lamport_keys
4. clean thread releases read lock
5. insert thread grabs write lock
6. insert thread inserts the 0-lamport update
7. State is bad since there is a zero lamport update for the key but it doesn't exist in the zero_lamport_keys set.